### PR TITLE
AP Scripting: Add health

### PR DESCRIPTION
**Pull request by haydendonald**
Added the ability to test sensor health through LUA.
This will be used for our internal testing which will use LUA.

### CODE REVIEW

File: libraries/AP_IOMCU/AP_IOMCU.h [hunk 1]
```diff
@@ -4,6 +4,7 @@
   For bootstrapping this will initially implement the px4io protocol,
   but will later move to an ArduPilot specific protocol
  */
+#pragma once
```
Comment (IamPete1) [inline libraries/AP_IOMCU/AP_IOMCU.h:7]: This change should be in a separate commit. `AP_IOMCU: ...`

File: libraries/AP_Scripting/generator/description/bindings.desc [hunk 2]
```diff
@@ -513,6 +514,9 @@ userdata AP_MotorsMatrix_Scripting_Dynamic::factor_table field throttle'array AP
 include AP_InertialSensor/AP_InertialSensor.h
 singleton AP_InertialSensor rename ins
 singleton AP_InertialSensor method get_temperature float uint8_t 0 INS_MAX_INSTANCES
+singleton AP_InertialSensor method healthy boolean
```
Comment (IamPete1) [inline libraries/AP_Scripting/generator/description/bindings.desc:517]: This function is just:
```
    bool healthy(void) const { return get_gyro_health() && get_accel_health(); }
```
So there is no need for the extra binding, you can do the same in lua as you have the other two bindings.

File: libraries/AP_Scripting/generator/description/bindings.desc [hunk 3]
```diff
@@ -628,6 +632,15 @@ singleton AP_Winch method release_length void float'skip_check
...
 singleton AP_Winch method get_rate_max float
 
+include AP_IOMCU/AP_IOMCU.h
+singleton AP_IOMCU depends HAL_WITH_IO_MCU
+singleton AP_IOMCU rename iomcu
+singleton AP_IOMCU method healthy boolean
+
+include AP_Compass/AP_Compass.h
+singleton Compass rename compass
+singleton Compass method healthy boolean uint8_t 0 COMPASS_MAX_INSTANCES
```
Comment (IamPete1) [inline libraries/AP_Scripting/generator/description/bindings.desc:642]: It would be nice to move away from limiting to `COMPASS_MAX_INSTANCES` but I think if you went outside that range bad things would happen.

The problem is that the user has no way to tell what `COMPASS_MAX_INSTANCES` is from the script, if they try and read a instance that is too high it will crash the script. Instead we would prefer to allow any index to be asked for and then return null if its invalid. 

There is the same issue on the `AP_InertialSensor` bindings of course, but your keeping in line with the existing pattern there.

File: libraries/AP_Scripting/generator/description/bindings.desc (hunk 2 above)
Comment (haydendonald) [reply libraries/AP_Scripting/generator/description/bindings.desc:517]: Ok, i added it for completion, but i'll remove it if you think that's best :)

File: libraries/AP_IOMCU/AP_IOMCU.h (hunk 1 above)
Comment (haydendonald) [reply libraries/AP_IOMCU/AP_IOMCU.h:7]: Ok, i'll move it over to a AP_IOMCU PR.

File: libraries/AP_Scripting/generator/description/bindings.desc (hunk 3 above)
Comment (haydendonald) [reply libraries/AP_Scripting/generator/description/bindings.desc:642]: Yeah, i didn't quite understand what would be best to put there, i copied how simular bindings were added. The other option that is available here is to add another binding for https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_Compass/AP_Compass.h#L156
```suggestion
singleton Compass method healthy boolean uint8_t 0 COMPASS_MAX_INSTANCES
singleton Compass method get_count uint8_t
```

File: libraries/AP_IOMCU/AP_IOMCU.h (hunk 1 above)
Comment (IamPete1) [reply libraries/AP_IOMCU/AP_IOMCU.h]: Same PR is fine, just needs to be in its own commit. We like to keep the commit history tidy with one per directory. 

https://ardupilot.org/dev/docs/git-interactive-rebase.html

Comment (haydendonald) [reply libraries/AP_IOMCU/AP_IOMCU.h:7]: Understood. I have made the change, first time using rebasing pretty cool stuff actually! :)

**Review by bugobliterator — APPROVED** [review]

**Review by IamPete1 — CHANGES_REQUESTED** [review]
Don't need singleton `AP_InertialSensor method healthy boolean` as this adds `get_gyro_health` and `get_accel_health`

**Comment by IamPete1** [comment]
500 bytes is quite a lot for this, it should all be checked as part of arming checks in anycase, I wonder if we could hook into those somehow.

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_Scripting/generator/description/bindings.desc (hunk 2 above)
Comment (haydendonald) [reply libraries/AP_Scripting/generator/description/bindings.desc]: This has been removed :)

File: libraries/AP_Scripting/generator/description/bindings.desc (hunk 3 above)
Comment (haydendonald) [reply libraries/AP_Scripting/generator/description/bindings.desc]: This should be resolved now https://github.com/ArduPilot/ardupilot/pull/23634/commits/9612dd8a4dee107ce8e630831b83a29e347c43ac :)

File: libraries/AP_Scripting/generator/description/bindings.desc [hunk 4]
```diff
@@ -334,6 +334,7 @@ singleton AP_Baro method get_pressure float
 singleton AP_Baro method get_temperature float
 singleton AP_Baro method get_external_temperature float
 singleton AP_Baro method get_altitude float
+singleton AP_Baro method healthy boolean uint8_t 0 BARO_MAX_INSTANCES
```
[the remark this answers is not in the document]
Comment (haydendonald) [reply libraries/AP_Scripting/generator/description/bindings.desc]: Done :) https://github.com/ArduPilot/ardupilot/pull/23634/commits/03314e895cededf788e201612fdad0b93febf551

**Review by IamPete1 — APPROVED** [review]
LGTM, Thanks.

File: libraries/AP_Baro/AP_Baro.cpp [hunk 5]
```diff
@@ -981,6 +981,17 @@ void AP_Baro::update(void)
...
 }
 
+#ifdef HAL_BUILD_AP_PERIPH
+// calibration and alt check not valid for AP_Periph
+bool AP_Baro::healthy(uint8_t instance) const {
+    return (instance < BARO_MAX_INSTANCES) ? sensors[instance].healthy : false;
+}
+#else
+bool AP_Baro::healthy(uint8_t instance) const {
+    return (instance < BARO_MAX_INSTANCES) ? (sensors[instance].healthy && sensors[instance].alt_ok && sensors[instance].calibrated) : false;
```
Comment (tridge) [inline libraries/AP_Baro/AP_Baro.cpp:991]: an if statement would be a lot clearer here if you don't mind

**Review by tridge — CHANGES_REQUESTED** [review]
looks good if you can fixup the baro change to be easier to read
thanks!

**Comment by tridge** [comment]
can be marged when baro function fixed up

File: libraries/AP_Baro/AP_Baro.cpp (hunk 5 above)
Comment (haydendonald) [reply libraries/AP_Baro/AP_Baro.cpp]: No worries! Done :)

**Comment by peterbarker** [comment]
Merged, thanks!
